### DISCUSSION TITLE
FAISS and Milvus Speed Benchmarking (Flat and HNSW)

### DISCUSSION THREAD

**Discussion by brandenchan** [milvus-io/milvus#4939] [Q&A and General discussion]
Hi Milvus community!

We at deepset.ai have been benchmarking the performance of FAISS against Milvus, in both the Flat and HNSW versions, in the hopes of releasing a blog post with these results (as discussed with @gujun720)! However, the results have not been quite as we expected. I have been using our search framework [Haystack](https://github.com/deepset-ai/haystack). More specifically, I am using [this benchmarking script](https://github.com/deepset-ai/haystack/blob/master/test/benchmarks/retriever.py).

When initializing the HNSW models, we have tried to align the parameters so they are comparable.

```
FAISS HNSW
n_links = 64
efSearch = 20
efConstruction = 80

Milvus HNSW
index_param = {"M": 64, "efConstruction": 80}
search_param = {"ef": 20}
```

The number below show queries per second given a certain document store, when loaded with a certain number of documents.

  | FAISS Flat | FAISS HNSW | Milvus Flat | Milvus HNSW
-- | -- | -- | -- | --
1000 | 40.41 | 44.32 | 40.48 | 40.39
10000 | 28.03 | 42.12 | 38.13 | 38.15
100000 | 6.62 | 40.68 | 28.30 | 28.08
500000 | 1.54 | 39.43 | 15.30 | 24.50

While Milvus Flat seems significantly faster than FAISS Flat, Milvus HNSW does not match the near constant speed that FAISS HNSW has. It is in fact only about as fast as Milvus Flat for 1k, 10k and 100k and is only faster at 500k. For reference, here are the mAP scores for the same configurations.

  | FAISS Flat | FAISS HNSW | Milvus Flat | Milvus HNSW
-- | -- | -- | -- | --
1000 | 92.95 | 92.95 | 92.95 | 92.95
10000 | 89.87 | 89.50 | 89.87 | 89.87
100000 | 86.55 | 84.33 | 86.55 | 86.55
500000 | 80.86 | 75.73 | 80.86 | 74.86

Here it shows that Milvus HNSW matches the Flat performance exactly at 1k, 10k and 100k. Only at 500k does the performance diverge.

So my questions are:
1) Should the Milvus numbers be always comparable / faster / slower than FAISS?
2) Did I initialize FAISS and Milvus HNSW correctly so that they can be directly compared?
3) How should HNSW speed scale as n_docs increases? Should it be near constant like FAISS HNSW is showing?
4) Do the mAP results give some clue as to what is happening? It seems up to 100k docs, Milvus HNSW is perhaps performing an exact NN search.

Thanks in advance!

**Comment by yhmo** [comment]
Hi @brandenchan

1. For hnsw idex, Milvus depends on [hnswlib](https://github.com/nmslib/hnswlib). The two library(hnswlib and faiss hnsw) may have difference on performance. From our manually testing, hnswlib is faster than faiss hnsw.
   For continually/frequently query request with small 'nq', milvus can combine multiple requests into one, to improve QPS.

2.The parameters is same, I think they can be compared correctly.
3. HNSW speed should be nearly constant. But as I mentioned, the Milvus can combine query request to improve QPS, so the QPS of milvus is better than FIASS
4. I am not sure about this. Is the mAP means recall rate?

**Reply by brandenchan** [reply]
Cool thanks @yhmo for the quick response, answers to questions 1, 2 and 3 all make sense to me.

As for the last one, mAP is mean average precision. It's a measure of how accurate the retrieval is. 

It seems that Milvus HNSW is running an exact NN search at 1k, 10k and 100k. But at 500k, it is both quicker and a little less accurate than Milvus Flat. Could this have something to do with how the parameters are configured and how the graph is constructed? Are there parameter values that make HNSW function like an exact NN search in cases where there are fewer documents?

**Reply by yhmo** [reply]
> Cool thanks @yhmo for the quick response, answers to questions 1, 2 and 3 all make sense to me.
> 
> As for the last one, mAP is mean average precision. It's a measure of how accurate the retrieval is.
> 
> It seems that Milvus HNSW is running an exact NN search at 1k, 10k and 100k. But at 500k, it is both quicker and a little less accurate than Milvus Flat. Could this have something to do with how the parameters are configured and how the graph is constructed? Are there parameter values that make HNSW function like an exact NN search in cases where there are fewer documents?

For this question, we are not sure now, we need to do some investigation.

**Comment by shengjun1985** [comment]
I tested the two HNSW on my local machine.

cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz * 8
data_set: SIFT1M (dim: 128, metric_type: L2)
M: 64, efConstruction: 80
ef: 20, nq: 1000,  topk: 10

【Faiss::IndexHNSWFlat】
| row count | build cost (ms) | query cost (ms) |
| ---------- | ---------------- | ------------ | 
| 1000     | 13 | 5 |
| 10000   | 138  | 7 |
| 100000 | 4899 | 30 |
| 500000 | 40044 | 42 |

【Milvus hnsw】
| row count | build cost (ms) | query cost (ms) |
| ---------- | ---------------- | ------------ | 
| 1000     | 15 | 3 |
| 10000   | 132  | 5 |
| 100000 | 4254 | 18 |
| 500000 | 34235 | 27 |

I didn't exactly calculate their recall rate. I was just looking at it roughly and Milvus hnsw's might be higher.
Is your test environment and parameters very different from mine?

**Reply by brandenchan** [reply]
Hi @shengjun1985 , thanks for those numbers. They don't quite match up to the numbers I have. I will double check my experimental run first and then get back to you.

**Comment by brandenchan** [comment]
Hi, we have started to revisit this topic but after a set of new experiments and some close inspection of our code, we were still unable to see any speedup in Milvus HNSW performance. Still it seems like it is identical to Milvus Flat until about 500k rows. 

@shengjun1985 As mentioned, the query cost numbers that you mention are significantly lower than what we were able to achieve. Would you be able to send us the script with which you ran your experiment? We'd like to compare initialization parameters.

We performed more experiments (see [here](https://github.com/deepset-ai/haystack/issues/968#issuecomment-842451263)) and found that varying Milvus HNSW's `M` parameter had no effect while varying the equivalent variable for FAISS HNSW did have an effect on both speed and accuracy (as expected). 

Here is [a simplified version of the script](https://github.com/deepset-ai/haystack/blob/milvus_simplified_script/test/benchmarks/simplified_script.py) that we are using in order to perform our benchmarking. 

We've looked very closely at everything but still can't see what could be going wrong. We are left with a few hypotheses:
1) The Milvus HNSW index we created have parameters such that it performs the same as Milvus Flat until we get to 500k+ rows (note this does not happen with FAISS HNSW)
2) The Milvus index we are initializing is in fact a Milvus Flat index (rather than HNSW) and the performance at 500k+ rows is an anomaly
3) Milvus' HNSW performance significantly differs from FAISS' HNSW performance. 

Would you be able to advise as to what we can try to get closer to a solution?

**Reply by yhmo** [reply]
Hi @brandenchan 

I just read your script. Seems the script is always using FLAT to search because the milvus index logic is special:
In milvus:
- A parameter named "index_file_size" is defined when collection is created, the default value is 1024MB. This value is a threshold for milvus automatically build index worker.
- Milvus split documents into segments according to the "index_file_size" value. For example, you totally insert 1.5GB data, there will be at least 2 segments in database, one is 1024MB+, the other is 500MB. You can refer this page: https://milvus.io/docs/v1.1.0/storage_concept.md
- There are two ways to create index for documents:
  1. create_collection() -> create_index() -> insert()
  2. create_collection() -> insert() -> create_index()
- For the way_1, if the inserted documents(vectors) accumulated to larger than 1024MB, the milvus will begin build index at the background thread. If there are more documents inserted, but not larger than 1024MB, the documents will keep in FLAT state. Search action will use index to search the index segment and brute-force search to the non-index segment. 
- For the way_2, since some documents have been inserted into the collection, after create_index(), milvus will build index for the documents even their size not large than 1024MB. Search action will use index to search all segments.

In the /haystack/test/benchmarks/simplified_script.py, it firstly initialize a MilvusDocumentStore object. In the MilvusDocumentStore::__init__(), _create_collection_and_index_if_not_exist() is called to invoke the milvus api create_collection() and create_index().
Then, MilvusDocumentStore::write_documents() is called to insert documents. 
Finally, a retriever call MilvusDocumentStore::query_by_embedding() to do search.
These steps is as same as the way_1, not all the documents are indexed, and the search performance is not best.

To make sure all segments are indexed before search, you can add a method "index_documents()" for the MilvusDocumentStore, then invoke this method after the write_documents():
```
    doc_store.write_documents(documents=docs, index=doc_index)
    doc_store.index_documents()
    doc_store.write_labels(labels=labels, index=label_index)
```
The index_documents() only do one thing:
```
def index_documents():
    status = self.milvus_server.create_index(self.index, self.index_type, self.index_param)
```

**Reply by brandenchan** [reply]
Thank you for pointing that out! By calling self.milvus_server.create_index() after document insertion, we were able to finally see the speedup that HNSW provides. Thanks for diving into our code and finding this issue!

Just one follow up question: Could you give some insight as to what might happen in other orders of operation? What's the best way to handle cases where a user might want to later insert more vectors? Is it very costly or problematic to call create_index() twice? For example:

i) create_collection() --> create_index() --> insert() --> create_index()
ii) create_collection() --> insert() --> create_index() --> insert() --> create_index()
iii) create_collection() --> create_index() --> insert() --> create_index() --> insert() --> create_index()

This would help us greatly in understanding how to implement Milvus best into our framework

**Reply by yhmo** [reply]
i) this way is better, the first create_index() tell milvus automatically build index for large segments, the second create_index() make sure build index for the "tail segment".
ii) this way is bad, no need to call create_index() after each insert(), because this will generate lot of small index
iii) bad, similar to ii)

**Reply by brandenchan** [reply]
Great thank you so much for your help!

**Reply by yhmo** [reply]
Just to confirm, the best way i) should be:
create_collection() --> create_index() --> insert() --> ...... -> insert() -> create_index()

**Reply by mataney** [reply]
Hi @yhmo 
Is this suggestion relevant only for HNSW or for all algorithms?

Thanks

**Comment by brandenchan** [comment]
So we reran our benchmarks after implementing the change that @yhmo suggested
```
create_collection() --> create_index() --> insert() --> create_index()` 
(Note that when we call `insert()` we insert all the documents that we have).
``` 
The results were significantly better and more in the range that we expect of an HNSW index. However, they seem to lag just a little behind FAISS HNSW in both speed and accuracy.

The results of our run using 500K vectors is here but the trend can be seen in at 100K, 10K and 1K as well. 

&nbsp; | FAISS Flat | FAISS HNSW | Milvus Flat | Milvus HNSW
-- | -- | -- | -- | --
Queries / second | 1.5 | 39.8 | 15.6 | 37.2
Milliseconds / Query | 650 | 25 | 64 | 27
Recall | 93.1 | 86.9 | 93.1 | 84.2
mAP | 80.9 | 75.7 | 80.9 | 73.6

I can understand if speed is perhaps a little off, but is it expected for recall and mAP to be different between FAISS and Milvus?

**Reply by shengjun1985** [reply]
I'm very confused why the recall rate of FLAT is not 100.
And how do you calculate the recall rate?

**Reply by yhmo** [reply]
@brandenchan 
I think the milvus hnsw should be faster than faiss hnsw.  I am not sure what's the problem. I will try to run the /haystack/test/benchmarks/simplified_script.py on my machine.

**Reply by yhmo** [reply]
After run the simplified_script.py, I got two reports:
```
{'date_time': datetime.datetime(2021, 6, 3, 11, 44, 12, 664911),
 'error': None,
 'map': 60.72872404830913,
 'n_queries': 5637,
 'queries_per_second': 6.059732725716486,
 'recall': 80.07805570338833,
 'retrieve_time': 930.2390476856381,
 'seconds_per_query': 0.1650237799690683,
 'top_k': 10}
........
{'date_time': datetime.datetime(2021, 6, 3, 14, 48, 16, 602645),
 'error': None,
 'map': 52.364690776079925,
 'n_queries': 5637,
 'queries_per_second': 15.890574584281442,
 'recall': 67.60688309384423,
 'retrieve_time': 354.73858859553,
 'seconds_per_query': 0.0629303864813784,
 'top_k': 10}
```

The first report is for flat, the second is for hnsw, both for milvus. But I don't know how to test FAISS by this script.

From the code, it seems the queries are executed one by one. Each test cost about 10 minutes on my machine. 
I think it can be faster. Milvus optimize QPS in request level, if there are multiple clients send query frequently/continually, milvus can combine small queries(query embedding is few) into large query. This behavior can improve QPS obviously. 
The example code execute 1000 * 10 queries by 10 threads, it will be faster than query one by one:
```
def search():
    for i in range(1000):
        embedding = get_embedding()
        status, result = milvus_server.search(collection_name=COLLECTION_NAME, top_k=TOPK,
                                       query_records=embedding, params=SEARCH_PARAM)

def multiple_queries():
    threads = []
    for k in range(10):
        x = threading.Thread(target=search, args=())
        threads.append(x)
        x.start()
    for th in threads:
        th.join()
```

**Reply by brandenchan** [reply]
Ok great! I will talk to the rest of the team about implementing this multiple query handling.

**Reply by brandenchan** [reply]
I've updated the script so that it has support for FAISS as well. You can find it [here](https://github.com/deepset-ai/haystack/blob/b42af2b06583ce857159a30649cc36741294953d/test/benchmarks/retriever_simplified.py). Note that it's been renamed and moved to a new branch

**Reply by yhmo** [reply]
Several points of investigation:
- In the **MilvusDocumentStore.query_by_embedding()**, no need to call **milvus_server.has_collection()**, since the search() will check the existence of the collection, if the collection doesn't exist, the serch() will return an error status. Remove the has_collection() can save 1 millisecond for each call.

- The report says each query cost 50+ milliseconds, but actually milvs only cost 2 ~ 3 milliseconds per query, I think the bottle neck is outside the query_by_embedding() method.

- I add some log in the /haystack/haystack/document_store/milvus.py and /haystack/haystack/document_store/faiss.py, to compare the real time cost of two servers. Here is the comparision:

milvus:
```
{'date_time': datetime.datetime(2021, 6, 4, 17, 5, 58, 700160),
 'error': None,
 'map': 52.08218234961196,
 'n_queries': 5637,
 'queries_per_second': 17.473708129090088,
 'recall': 67.12790491396132,
 'retrieve_time': 322.59895600611344,
 'seconds_per_query': 0.057228837325902684,
 'top_k': 10}

......
06/04/2021 17:05:58 - INFO - haystack.document_store.milvus -   ======= milvus search time cost: 2.7582645416259766 ms
06/04/2021 17:05:58 - INFO - haystack.document_store.milvus -   ======= milvus search time cost: 2.707958221435547 ms
06/04/2021 17:05:58 - INFO - haystack.document_store.milvus -   ======= milvus search total cost: 16433.369636535645 ms
```

faiss:
```
{'date_time': datetime.datetime(2021, 6, 4, 16, 55, 34, 937137),
 'error': None,
 'map': 55.81270714187171,
 'n_queries': 5637,
 'queries_per_second': 18.41110687142562,
 'recall': 73.05304239843889,
 'retrieve_time': 306.17387859220617,
 'seconds_per_query': 0.054315039665106646,
 'top_k': 10}

......
06/04/2021 16:55:34 - INFO - haystack.document_store.faiss -   ======= faiss search time cost: 1.5094280242919922 ms
06/04/2021 16:55:34 - INFO - haystack.document_store.faiss -   ======= faiss search time cost: 1.6372203826904297 ms
06/04/2021 16:55:34 - INFO - haystack.document_store.faiss -   ======= faiss search total cost: 8771.78978919983 ms
```

From the report, we can see each query cost 50+ milliseconds, but in log, milvus search() only cost 2 ~ 3 ms per query, faiss only cost 1 ~ 2 ms. Milvus cost more time than faiss, this is because milvus waste about 1 ms on rpc transfer/encode/decode.

- The difference of mAP, I think is caused by different lib. Milvus use the https://github.com/nmslib/hnswlib for hnsw index.

**Comment by brandenchan** [comment]
As an aside I just want to say that we found that Milvus Flat was significantly better than FAISS Flat in indexing and querying speeds so thank you very much for your work! 

Happy to share more of our results if they are of interest to you. We will also be publishing an article soon on the Haystack-Milvus integration.

**Reply by yunjiangster** [reply]
@brandenchan do you have some intuition why Milvus Flat is so much faster than FAISS Flat?
